### ISSUE TITLE
Upgrade fails from 6.6.0 to 6.6.2+

### ISSUE THREAD

**Issue by tareko** [mattermost/mattermost-server#20580]
Per Mattermost guidelines, GitHub issues are for bug reports: <https://handbook.mattermost.com/contributors/contributors/ways-to-contribute/>.

For troubleshooting see: https://forum.mattermost.com/.
For feature proposals see: https://www.mattermost.com/feature-ideas/

If you've found a bug--something appears unintentional--please follow these steps:

1. Confirm you’re filing a new issue. [Search existing tickets in Jira](https://mattermost.atlassian.net/jira/software/c/projects/MM/issues/) to ensure that the ticket does not already exist.
2. Confirm your issue does not involve security. Otherwise, please see our [Responsible Disclosure Policy](https://mattermost.com/security-vulnerability-report/).
3. [File a new issue](https://github.com/mattermost/mattermost-server/issues/new) using the format below. Mattermost will confirm steps to reproduce and file in Jira, or ask for more details if there is trouble reproducing it. If there's already an existing bug in Jira, it will be linked back to the GitHub issue so you can track when it gets fixed. 

#### Summary
Database upgrade fails from 6.6.0 to 6.7.0 or above.

#### Steps to reproduce
Start with Mattermost 6.6.0. Apply upgrade to 6.7.0 or above by restarting server.

#### Expected behavior
When upgrading, server should be able to apply migrations successfully from 6.6.0 onward.

#### Observed behavior (that appears unintentional) 
When upgrading from 6.6.0 to 6.7.0 or above, i get:

```
2022/07/03 02:24:25 ==  update_configurations_sha: migrating  ==========================================================
Error: failed to load configuration: failed to initialize: driver: mysql, message: failed when applying migration, command: apply_migration, originalError: Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '""' at line 1, query: 

SET @preparedStatement = (SELECT IF(
    (
        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
        WHERE table_name = 'Configurations'
        AND table_schema = DATABASE()
        AND column_name = 'SHA'
    ) > 0,
    'SELECT 1',
    'ALTER TABLE Configurations ADD COLUMN SHA char(64) DEFAULT "";'
));

PREPARE alterIfNotExists FROM @preparedStatement;
EXECUTE alterIfNotExists;
DEALLOCATE PREPARE alterIfNotExists;


Usage:
  mattermost [flags]
  mattermost [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  db          Commands related to the database
  export      Export data from Mattermost
  help        Help about any command
  import      Import data.
  jobserver   Start the Mattermost job server
  server      Run the Mattermost server
  version     Display version information

Flags:
  -c, --config string   Configuration file to use.
  -h, --help            help for mattermost

Use "mattermost [command] --help" for more information about a command.
```

When attempting to upgrade from 6.6.0 to 6.6.2, I get the following:

```
{"timestamp":"2022-07-03 03:52:21.064 Z","level":"fatal","msg":"Failed to apply database migrations.","caller":"sqlstore/store.go:160","error":"driver: mysql, message: failed when applying migration, command: apply_migration, originalError: Error 1054: Unknown column '' in 'field list', query: \n\nSET @preparedStatement = (SELECT IF(\n    (\n        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS\n        WHERE table_name = 'OAuthApps'\n        AND table_schema = DATABASE()\n        AND column_name = 'MattermostAppID'\n    ) > 0,\n    'UPDATE OAuthApps SET MattermostAppID = \"\" WHERE MattermostAppID IS NULL;',\n    'SELECT 1'\n));\n\nPREPARE alterIfExists FROM @preparedStatement;\nEXECUTE alterIfExists;\nDEALLOCATE PREPARE alterIfExists;\n\nSET @preparedStatement = (SELECT IF(\n    (\n        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS\n        WHERE table_name = 'OAuthApps'\n        AND table_schema = DATABASE()\n        AND column_name = 'MattermostAppID'\n    ) > 0,\n    'ALTER TABLE OAuthApps MODIFY MattermostAppID varchar(32) NOT NULL DEFAULT \"\";',\n    'SELECT 1'\n));\n\nPREPARE alterIfExists FROM @preparedStatement;\nEXECUTE alterIfExists;\nDEALLOCATE PREPARE alterIfExists;\n\n"}
```

#### Possible fixes
If you can, link to the line of code that might be responsible for the problem

I am using Digital Ocean Kubernetes with a managed database (MySQL 8.0.28). After reading Issue #19602 , I also gave full permissions to the mattermost user and tried switching to the administrative user to rule out permissions issues - neither worked.

**Comment by tareko** [comment]
## Possible fix:

The commands that are executed seem to be run with `""`. However, `''` seems to solve this, albeit with updating zero rows in my installation.

**Comment by streamer45** [comment]
@isacikgoz Would you be able to help with this?

**Comment by isacikgoz** [comment]
Hey @tareko the reason that we used double quotes that we define prepared statements in single quotes. It didn't raised any problem on my setup though (both 5.7 and 8). Maybe we can change those to escaped single quotes ie `\'`. WDYT @agnivade?

**Comment by agnivade** [comment]
Hmm, I think we need to first understand why this is failing. If you cannot reproduce it in MySQL 8, maybe the problem is elsewhere? Is your version 8.0.28 specifically? We've had some JSON bugs in dot releases before, so sometimes the dot release version matters as well.

**Comment by isacikgoz** [comment]
Yes, it I tested with the exact same mysql 8 version (`v8.0.28` docker image). I am not so familiar with DO managed mysql though, if they have different configuration.

**Comment by agnivade** [comment]
That sounds good and would make sense. Is there a config option somewhere that controls single quotes vs double quotes? :thinking:

**Comment by tareko** [comment]
I'm very happy to create an account or make a separate installation for you to experiment with. If yes, let me know and I'll dm

**Comment by tareko** [comment]
Thank you. I think this helped me track the problem down.

Here is the output:

```
REAL_AS_FLOAT,PIPES_AS_CONCAT,ANSI_QUOTES,IGNORE_SPACE,ONLY_FULL_GROUP_BY,ANSI,STRICT_ALL_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION
```

I was curious about `ANSI_QUOTES`, and when I looked it up, this is what the [MySQL dev page](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_ansi_quotes) had to say:

```
ANSI_QUOTES

Treat " as an identifier quote character (like the ` quote character) and not as a string quote character. You can still use ` to quote identifiers with this mode enabled. With ANSI_QUOTES enabled, you cannot use double quotation marks to quote literal strings because they are interpreted as identifiers.
```

I solved this by going to the database settings and deleting this mode:

![Screenshot from 2022-07-11 07-32-23](https://user-images.githubusercontent.com/1128840/178255593-403395b0-b4fc-4ae4-a91f-db3f5fb53435.png)

This appeared to have fixed my problem. I wonder if this could be noted in the requirements or checked for by the installer / updater?

**Comment by agnivade** [comment]
Thanks @tareko. I agree, we can do that. Are you interested in submitting a fix? I can give you some pointers as to how to do this.

**Comment by tareko** [comment]
I'm happy to, but not sure how / where. Do you mean documentation? Or something more substantial?

**Comment by agnivade** [comment]
I meant in the code, when we initialize our sql store. The code is here: https://github.com/mattermost/mattermost-server/blob/5f4da3f3082e6aa5323cc31e3cbf99a58281239c/store/sqlstore/store.go#L153. We do several checks to ensure minimum DB version, collation etc. We just need to add another similar check and ensure that `ANSI_QUOTES` isn't there.

**Comment by hanzei** [comment]
After multiple years of inactivity, we're closing this to help keep our issue tracker focused on active work.
